README.md
# k8scontroller-sentinel

![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/vfabi/k8scontroller-sentinel)
![GitHub last commit](https://img.shields.io/github/last-commit/vfabi/k8scontroller-sentinel)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

[![Generic badge](https://img.shields.io/badge/hub.docker.com-vfabi/k8scontroller_sentinel-<>.svg)](https://hub.docker.com/repository/docker/vfabi/k8scontroller-sentinel)
![Docker Pulls](https://img.shields.io/docker/pulls/vfabi/k8scontroller-sentinel)
![Docker Cloud Automated build](https://img.shields.io/docker/cloud/automated/vfabi/k8scontroller-sentinel)
![Docker Cloud Build Status](https://img.shields.io/docker/cloud/build/vfabi/k8scontroller-sentinel)

Kubernetes controller to manage some special tasks.

## Status

Beta

## Features

- handler to remove pods with status phase=Failed and reason=Shutdown
- helper method to send messages to Slack

## Technology stack

- Python 3.8+
- KOPF - Kubernetes OPerators Framework

## Configuration

### Environment variables

Name | Description | Mandatory | Default | Example
--- | --- | --- | --- | ---
SLACK_URL | Slack incoming web hook URL. For sending message to Slack. | true | | https://hooks.slack.com/services/T127SJ7LZJ5/A12D22SFPD3/11gEI9FlguRBPvLVkydtDs22k |
SLACK_CHANNEL | Slack channel. For sending message to Slack. | true | | #alerts |

## Usage

### Deployment

1. Check configuration: environment variables in `deployment/kubernetes/configmap.yaml`
2. Just apply all yaml files from `deployment/kubernetes/` folder.

## Docker

[![Generic badge](https://img.shields.io/badge/hub.docker.com-vfabi/k8scontroller_sentinel-<>.svg)](https://hub.docker.com/repository/docker/vfabi/k8scontroller-sentinel)

## Contributing

Please refer to each project's style and contribution guidelines for submitting patches and additions. In general, we follow the "fork-and-pull" Git workflow.

 1. **Fork** the repo on GitHub
 2. **Clone** the project to your own machine
 3. **Commit** changes to your own branch
 4. **Push** your work back up to your fork
 5. Submit a **Pull request** so that we can review your changes

NOTE: Be sure to merge the latest from "upstream" before making a pull request!

## License

Apache 2.0


CHANGELOG.md
# CHANGELOG.md

## 0.0.1

Features:

- add base functionality


app/helpers.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
    @project: k8scontroller-sentinel
    @copyright: © 2021 by vfabi
    @author: vfabi
    @support: vfabi
    @initial date: 2021-09-05 11:23:27
    @license: this file is subject to the terms and conditions defined
        in file 'LICENSE.txt', which is part of this source code package
    @description:
    @todo:
"""

import json
import sys
import requests


class Sender:

    def __init__(self, channel_type='slack', **kwargs):
        if channel_type == 'slack':
            self.channel_type = 'slack'
            self.slack_url = kwargs['slack_url']
            self.slack_channel = kwargs['slack_channel']
        else:
            raise Exception('This channel_type is not implemented.')

    def sendSlack(self, message):
        url = self.slack_url
        message = (message)
        title = ('kubernetes event :zap:')
        slack_data = {
            'username': 'k8scontroller-sentinel',
            'icon_emoji': ':jack_o_lantern:',
            'channel': self.slack_channel,
            'attachments': [
                {
                    'color': '#9733EE',
                    'fields': [
                        {
                            'title': title,
                            'value': message,
                            'short': 'false',
                        }
                    ]
                }
            ]
        }
        byte_length = str(sys.getsizeof(slack_data))
        headers = {'Content-Type': 'application/json', 'Content-Length': byte_length}
        response = requests.post(url, data=json.dumps(slack_data), headers=headers)
        if response.status_code != 200:
            raise Exception(response.status_code, response.text)

    def send(self, message):
        if self.channel_type == 'slack':
            self.sendSlack(message)


app/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
    @project: k8scontroller-sentinel
    @copyright: © 2021 by vfabi
    @author: vfabi
    @support: vfabi
    @initial date: 2021-09-05 11:23:27
    @license: this file is subject to the terms and conditions defined
        in file 'LICENSE.txt', which is part of this source code package
    @description:
    @todo:
"""

import os
import asyncio
import logging
import kopf
import pykube
from pykube.exceptions import ObjectDoesNotExist
from helpers import Sender

SLACK_URL = os.getenv('SLACK_URL')
SLACK_CHANNEL = os.getenv('SLACK_CHANNEL')

logger = logging.getLogger()


@kopf.on.startup()
def configure(settings: kopf.OperatorSettings, **_):
    settings.posting.level = logging.INFO
    settings.watching.connect_timeout = 1 * 60
    settings.watching.server_timeout = 10 * 60
    # settings.posting.enabled = False


@kopf.on.event('', 'v1', 'pods')
def event(event, **kwargs):
    sender = Sender(
        channel_type='slack',
        slack_url=SLACK_URL,
        slack_channel=SLACK_CHANNEL
    )
    api = pykube.HTTPClient(pykube.KubeConfig.from_service_account())

    try:
        namespace = event['object']['metadata']['namespace']
        pod = pykube.Pod.objects(api, namespace=namespace).get_by_name(event['object']['metadata']['name'])
        pod_status_phase = pod.obj['status']['phase']
        if pod_status_phase == 'Failed':
            pod_status_reason = pod.obj['status']['reason']
        else:
            pod_status_reason = None
        logger.info(f'Got pod event. Pod={namespace}/{pod}, phase={pod_status_phase}, reason={pod_status_reason}.')

        if pod_status_reason == 'Shutdown':
            logger.warning(f'Deleting pod {namespace}/{pod}, phase={pod_status_phase}, reason={pod_status_reason}, ready={pod.ready}.')
            pod.delete()
            sender.send(f'*pod:* {namespace}/{pod}\n *status:* {pod_status_phase}\n *reason:* {pod_status_reason}\n *action:* deleted')
    except ObjectDoesNotExist as e:
        logger.warning(f'Pod does not exists. Details: {e}')
    except Exception as e:
        logger.error(f'Exception. Details: {e}')

    api.session.close()


@kopf.on.login()
def login(**kwargs):
    return kopf.login_with_service_account(**kwargs)


def main():
    asyncio.run(kopf.operator(
        # registry=kopf.OperatorRegistry(),
        clusterwide=True,
        standalone=True,
    ))


main()
